### ISSUE TITLE
./build_rpanion.sh can't get npm installation complete in PiZeroW/2022-09-22-raspios-bullseye-armhf-lite

### ISSUE THREAD

**Issue by lida2003** [stephendade/Rpanion-server#192]
commid id: [3bd162b](https://github.com/stephendade/Rpanion-server/commit/3bd162b610fd52dbcd947ac19511ab27127ab8fd)

- 2022-09-22-raspios-bullseye-armhf-lite.img.xz + PiZeroW ==> Fail
- 2022-09-22-raspios-bullseye-arm64-lite.img.xz + Pi3B+ ==> OK

**First time install, it's connected to internet.**
```
+ cd ../../deploy
+ ./build_rpanion.sh
+ cd ../
++ free -m
++ awk '/^Mem:/{print $2}'
+ '[' 429 -le 500 ']'
+ export NODE_OPTIONS=--max-old-space-size=256
+ NODE_OPTIONS=--max-old-space-size=256
+ MAX_RETRIES=3
+ NPM_INSTALL_EXIT_CODE=0
++ seq 1 3
+ for i in $(seq 1 $MAX_RETRIES)
+ echo 'Running npm install (attempt 1)...'
Running npm install (attempt 1)...
+ npm install
npm ERR! code ETIMEDOUT
npm ERR! syscall connect
npm ERR! errno ETIMEDOUT
npm ERR! network request to https://registry.npmjs.org/npm/-/npm-9.8.1.tgz failed, reason: connect ETIMEDOUT 104.16.0.35:443
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/daniel/.npm/_logs/2024-01-14T23_51_35_034Z-debug-0.log



$ cat /home/daniel/.npm/_logs/2024-01-14T23_51_35_034Z-debug-0.log
0 verbose cli /usr/local/lib/nodejs/node-v16.19.1-linux-armv6l/bin/node /usr/local/bin/npm
1 info using npm@8.19.3
2 info using node@v16.19.1
3 timing npm:load:whichnode Completed in 3ms
4 timing config:load:defaults Completed in 47ms
5 timing config:load:file:/usr/local/lib/nodejs/node-v16.19.1-linux-armv6l/lib/node_modules/npm/npmrc Completed in 14ms
6 timing config:load:builtin Completed in 18ms
7 timing config:load:cli Completed in 38ms
8 timing config:load:env Completed in 7ms
9 timing config:load:file:/home/daniel/Rpanion-server/.npmrc Completed in 1ms
10 timing config:load:project Completed in 24ms
11 timing config:load:file:/home/daniel/.npmrc Completed in 1ms
12 timing config:load:user Completed in 6ms
13 timing config:load:file:/usr/local/lib/nodejs/node-v16.19.1-linux-armv6l/etc/npmrc Completed in 2ms
14 timing config:load:global Completed in 4ms
15 timing config:load:validate Completed in 2ms
16 timing config:load:credentials Completed in 27ms
17 timing config:load:setEnvs Completed in 20ms
18 timing config:load Completed in 207ms
19 timing npm:load:configload Completed in 211ms
20 timing npm:load:mkdirpcache Completed in 23ms
21 timing npm:load:mkdirplogs Completed in 25ms
22 verbose title npm install
23 verbose argv "install"
24 timing npm:load:setTitle Completed in 20ms
25 timing config:load:flatten Completed in 68ms
26 timing npm:load:display Completed in 112ms
27 verbose logfile logs-max:10 dir:/home/daniel/.npm/_logs
28 verbose logfile /home/daniel/.npm/_logs/2024-01-14T23_51_35_034Z-debug-0.log
29 timing npm:load:logFile Completed in 113ms
30 timing npm:load:timers Completed in 2ms
31 timing npm:load:configScope Completed in 1ms
32 timing npm:load Completed in 759ms
33 timing arborist:ctor Completed in 17ms
34 silly logfile done cleaning log files
35 timing idealTree:init Completed in 47171ms
36 timing idealTree:userRequests Completed in 4ms
37 silly idealTree buildDeps
38 timing idealTree:#root Completed in 11ms
39 timing idealTree:buildDeps Completed in 36ms
40 timing idealTree:fixDepFlags Completed in 8ms
41 timing idealTree Completed in 48133ms
42 timing reify:loadTrees Completed in 49079ms
43 timing reify:diffTrees Completed in 1038ms
44 silly reify moves {}
45 timing reify:retireShallow Completed in 26ms
46 timing reify:createSparse Completed in 13971ms
47 silly tarball no local data for npm@https://registry.npmjs.org/npm/-/npm-9.8.1.tgz. Extracting by manifest.
48 timing reify:rollback:createSparse Completed in 8603ms
49 timing reify:rollback:retireShallow Completed in 1ms
50 timing command:install Completed in 538910ms
51 verbose type system
52 verbose stack FetchError: request to https://registry.npmjs.org/npm/-/npm-9.8.1.tgz failed, reason: connect ETIMEDOUT 104.16.0.35:443
52 verbose stack     at ClientRequest.<anonymous> (/usr/local/lib/nodejs/node-v16.19.1-linux-armv6l/lib/node_modules/npm/node_modules/minipass-fetch/lib/index.js:130:14)
52 verbose stack     at ClientRequest.emit (node:events:513:28)
52 verbose stack     at TLSSocket.socketErrorListener (node:_http_client:494:9)
52 verbose stack     at TLSSocket.emit (node:events:525:35)
52 verbose stack     at emitErrorNT (node:internal/streams/destroy:157:8)
52 verbose stack     at emitErrorCloseNT (node:internal/streams/destroy:122:3)
52 verbose stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
53 verbose cwd /home/daniel/Rpanion-server
54 verbose Linux 6.1.21+
55 verbose node v16.19.1
56 verbose npm  v8.19.3
57 error code ETIMEDOUT
58 error syscall connect
59 error errno ETIMEDOUT
60 error network request to https://registry.npmjs.org/npm/-/npm-9.8.1.tgz failed, reason: connect ETIMEDOUT 104.16.0.35:443
61 error network This is a problem related to network connectivity.
61 error network In most cases you are behind a proxy or have bad network settings.
61 error network
61 error network If you are behind a proxy, please make sure that the
61 error network 'proxy' config is set properly.  See: 'npm help config'
62 verbose exit 1
63 timing npm Completed in 542136ms
64 verbose unfinished npm timer reify 1705276298136
65 verbose unfinished npm timer reify:loadBundles 1705276363232
66 verbose unfinished npm timer reifyNode:node_modules/npm 1705276363245
67 verbose code 1
68 error A complete log of this run can be found in:
68 error     /home/daniel/.npm/_logs/2024-01-14T23_51_35_034Z-debug-0.log
```

**second time install, add a proxy, still no luck**
```
$ export https_proxy=http://192.168.1.12:808/
export http_proxy=http://192.168.1.12:808/
daniel@raspberrypi:~/Rpanion-server/deploy $ ./build_rpanion.sh
+ cd ../
++ awk '/^Mem:/{print $2}'
++ free -m
+ '[' 429 -le 500 ']'
+ export NODE_OPTIONS=--max-old-space-size=256
+ NODE_OPTIONS=--max-old-space-size=256
+ MAX_RETRIES=3
+ NPM_INSTALL_EXIT_CODE=0
++ seq 1 3
+ for i in $(seq 1 $MAX_RETRIES)
+ echo 'Running npm install (attempt 1)...'
Running npm install (attempt 1)...
+ npm install
(#########⠂⠂⠂⠂⠂⠂⠂⠂⠂) ⠇ reify:npm: timing reify:createSparse Completed in 13944ms
(#########⠂⠂⠂⠂⠂⠂⠂⠂⠂) ⠇ reify:npm: timing reify:createSparse Completed in 13944ms
npm ERR! code ECONNRESET
npm ERR! network aborted
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/daniel/.npm/_logs/2024-01-15T00_06_24_857Z-debug-0.log
```

**third time, npm install fail**
```
daniel@raspberrypi:~/Rpanion-server/deploy $ npm install
npm ERR! code ECONNRESET
npm ERR! network aborted
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/daniel/.npm/_logs/2024-01-15T00_23_02_624Z-debug-0.log
daniel@raspberrypi:~/Rpanion-server/deploy $ echo $?
1
```

**fourth time, build again, the npm install loop doesn't work as expected**
```
$ ./build_rpanion.sh
+ cd ../
++ free -m
++ awk '/^Mem:/{print $2}'
+ '[' 429 -le 500 ']'
+ export NODE_OPTIONS=--max-old-space-size=256
+ NODE_OPTIONS=--max-old-space-size=256
+ MAX_RETRIES=3
+ NPM_INSTALL_EXIT_CODE=0
++ seq 1 3
+ for i in $(seq 1 $MAX_RETRIES)
+ echo 'Running npm install (attempt 1)...'
Running npm install (attempt 1)...
+ npm install
npm ERR! code ENOTEMPTY
npm ERR! syscall rename
npm ERR! path /home/daniel/Rpanion-server/node_modules/acorn-globals
npm ERR! dest /home/daniel/Rpanion-server/node_modules/.acorn-globals-MiwWVQlT
npm ERR! errno -39
npm ERR! ENOTEMPTY: directory not empty, rename '/home/daniel/Rpanion-server/node_modules/acorn-globals' -> '/home/daniel/Rpanion-server/node_modules/.acorn-globals-MiwWVQlT'

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/daniel/.npm/_logs/2024-01-15T00_37_32_731Z-debug-0.log
```

**Comment by lida2003** [comment]
It has been sucessfully depolyed once. Failed anyway.

- 2023-02-21-raspios-bullseye-armhf-lite.img.xz +  + PiZeroW ==> Fail

```
+ cd ../../deploy
+ ./build_rpanion.sh
+ cd ../
++ awk '/^Mem:/{print $2}'
++ free -m
+ '[' 429 -le 500 ']'
+ export NODE_OPTIONS=--max-old-space-size=256
+ NODE_OPTIONS=--max-old-space-size=256
+ MAX_RETRIES=3
+ NPM_INSTALL_EXIT_CODE=0
++ seq 1 3
+ for i in $(seq 1 $MAX_RETRIES)
+ echo 'Running npm install (attempt 1)...'
Running npm install (attempt 1)...
+ npm install
(#########⠂⠂⠂⠂⠂⠂⠂⠂⠂) ⠇ reify:npm: timing reify:createSparse Completed in 14442ms
npm ERR! code ETIMEDOUT
npm ERR! syscall connect
npm ERR! errno ETIMEDOUT
npm ERR! network request to https://registry.npmjs.org/npm/-/npm-9.8.1.tgz failed, reason: connect ETIMEDOUT 104.16.24.34:443
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/daniel/.npm/_logs/2024-01-15T02_11_39_592Z-debug-0.log


$ cat /home/daniel/.npm/_logs/2024-01-15T02_11_39_592Z-debug-0.log
0 verbose cli /usr/local/lib/nodejs/node-v16.19.1-linux-armv6l/bin/node /usr/local/bin/npm
1 info using npm@8.19.3
2 info using node@v16.19.1
3 timing npm:load:whichnode Completed in 3ms
4 timing config:load:defaults Completed in 47ms
5 timing config:load:file:/usr/local/lib/nodejs/node-v16.19.1-linux-armv6l/lib/node_modules/npm/npmrc Completed in 14ms
6 timing config:load:builtin Completed in 18ms
7 timing config:load:cli Completed in 39ms
8 timing config:load:env Completed in 7ms
9 timing config:load:file:/home/daniel/Rpanion-server/.npmrc Completed in 1ms
10 timing config:load:project Completed in 25ms
11 timing config:load:file:/home/daniel/.npmrc Completed in 2ms
12 timing config:load:user Completed in 6ms
13 timing config:load:file:/usr/local/lib/nodejs/node-v16.19.1-linux-armv6l/etc/npmrc Completed in 2ms
14 timing config:load:global Completed in 4ms
15 timing config:load:validate Completed in 3ms
16 timing config:load:credentials Completed in 27ms
17 timing config:load:setEnvs Completed in 20ms
18 timing config:load Completed in 208ms
19 timing npm:load:configload Completed in 212ms
20 timing npm:load:mkdirpcache Completed in 22ms
21 timing npm:load:mkdirplogs Completed in 10ms
22 verbose title npm install
23 verbose argv "install"
24 timing npm:load:setTitle Completed in 19ms
25 timing config:load:flatten Completed in 69ms
26 timing npm:load:display Completed in 112ms
27 verbose logfile logs-max:10 dir:/home/daniel/.npm/_logs
28 verbose logfile /home/daniel/.npm/_logs/2024-01-15T02_11_39_592Z-debug-0.log
29 timing npm:load:logFile Completed in 113ms
30 timing npm:load:timers Completed in 1ms
31 timing npm:load:configScope Completed in 1ms
32 timing npm:load Completed in 513ms
33 timing arborist:ctor Completed in 22ms
34 silly logfile done cleaning log files
35 timing idealTree:init Completed in 47738ms
36 timing idealTree:userRequests Completed in 4ms
37 silly idealTree buildDeps
38 timing idealTree:#root Completed in 11ms
39 timing idealTree:buildDeps Completed in 36ms
40 timing idealTree:fixDepFlags Completed in 10ms
41 timing idealTree Completed in 48726ms
42 timing reify:loadTrees Completed in 49033ms
43 timing reify:diffTrees Completed in 1419ms
44 silly reify moves {}
45 timing reify:retireShallow Completed in 27ms
46 timing reify:createSparse Completed in 14442ms
47 silly tarball no local data for npm@https://registry.npmjs.org/npm/-/npm-9.8.1.tgz. Extracting by manifest.
48 timing reify:rollback:createSparse Completed in 8956ms
49 timing reify:rollback:retireShallow Completed in 1ms
50 timing command:install Completed in 539375ms
51 verbose type system
52 verbose stack FetchError: request to https://registry.npmjs.org/npm/-/npm-9.8.1.tgz failed, reason: connect ETIMEDOUT 104.16.24.34:443
52 verbose stack     at ClientRequest.<anonymous> (/usr/local/lib/nodejs/node-v16.19.1-linux-armv6l/lib/node_modules/npm/node_modules/minipass-fetch/lib/index.js:130:14)
52 verbose stack     at ClientRequest.emit (node:events:513:28)
52 verbose stack     at TLSSocket.socketErrorListener (node:_http_client:494:9)
52 verbose stack     at TLSSocket.emit (node:events:525:35)
52 verbose stack     at emitErrorNT (node:internal/streams/destroy:157:8)
52 verbose stack     at emitErrorCloseNT (node:internal/streams/destroy:122:3)
52 verbose stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
53 verbose cwd /home/daniel/Rpanion-server
54 verbose Linux 5.15.84+
55 verbose node v16.19.1
56 verbose npm  v8.19.3
57 error code ETIMEDOUT
58 error syscall connect
59 error errno ETIMEDOUT
60 error network request to https://registry.npmjs.org/npm/-/npm-9.8.1.tgz failed, reason: connect ETIMEDOUT 104.16.24.34:443
61 error network This is a problem related to network connectivity.
61 error network In most cases you are behind a proxy or have bad network settings.
61 error network
61 error network If you are behind a proxy, please make sure that the
61 error network 'proxy' config is set properly.  See: 'npm help config'
62 verbose exit 1
63 timing npm Completed in 542288ms
64 verbose unfinished npm timer reify 1705284702419
65 verbose unfinished npm timer reify:loadBundles 1705284768302
66 verbose unfinished npm timer reifyNode:node_modules/npm 1705284768305
67 verbose code 1
68 error A complete log of this run can be found in:
68 error     /home/daniel/.npm/_logs/2024-01-15T02_11_39_592Z-debug-0.log

```

**Comment by lida2003** [comment]
Tried another npm source.

```
+ npm install
npm WARN deprecated w3c-hr-time@1.0.2: Use your platform's native performance.now() and performance.timeOrigin.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated express-pino-logger@7.0.0: use pino-http instead
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.

changed 1063 packages in 19m

304 packages are looking for funding
  run `npm fund` for details
+ NPM_INSTALL_EXIT_CODE=0
+ '[' 0 -eq 0 ']'
+ echo 'npm install succeeded on attempt 1'
npm install succeeded on attempt 1
+ break
+ npm run build

> Rpanion-server@0.10.0 build
> DISABLE_ESLINT_PLUGIN=true react-scripts build

node:internal/modules/cjs/loader:433
      throw err;
      ^

Error: Cannot find module '/home/daniel/Rpanion-server/node_modules/tryer/src/tryer'. Please verify that the package.json has a valid "main" entry
    at tryPackage (node:internal/modules/cjs/loader:425:19)
    at Function.Module._findPath (node:internal/modules/cjs/loader:638:18)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1007:27)
    at Function.Module._load (node:internal/modules/cjs/loader:866:27)
    at Module.require (node:internal/modules/cjs/loader:1093:19)
    at require (node:internal/modules/cjs/helpers:108:18)
    at Object.<anonymous> (/home/daniel/Rpanion-server/node_modules/bfj/src/streamify.js:9:15)
    at Module._compile (node:internal/modules/cjs/loader:1191:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1245:10)
    at Module.load (node:internal/modules/cjs/loader:1069:32) {
  code: 'MODULE_NOT_FOUND',
  path: '/home/daniel/Rpanion-server/node_modules/tryer/package.json',
  requestPath: 'tryer'
}

```

**Comment by lida2003** [comment]
Hmm.... still trouble. 

I have tried "npm config set registry https://registry.npm.taobao.org/", then I got below error.

```
+ echo 'Running npm install (attempt 1)...'
Running npm install (attempt 1)...
+ npm install

up to date in 5m

304 packages are looking for funding
  run `npm fund` for details
+ NPM_INSTALL_EXIT_CODE=0
+ '[' 0 -eq 0 ']'
+ echo 'npm install succeeded on attempt 1'
npm install succeeded on attempt 1
+ break
+ npm run build

> Rpanion-server@0.10.0 build
> DISABLE_ESLINT_PLUGIN=true react-scripts build

node:internal/modules/cjs/loader:433
      throw err;
      ^

Error: Cannot find module '/home/daniel/Rpanion-server/node_modules/tryer/src/tryer'. Please verify that the package.json has a valid "main" entry
    at tryPackage (node:internal/modules/cjs/loader:425:19)
    at Function.Module._findPath (node:internal/modules/cjs/loader:638:18)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1007:27)
    at Function.Module._load (node:internal/modules/cjs/loader:866:27)
    at Module.require (node:internal/modules/cjs/loader:1093:19)
    at require (node:internal/modules/cjs/helpers:108:18)
    at Object.<anonymous> (/home/daniel/Rpanion-server/node_modules/bfj/src/streamify.js:9:15)
    at Module._compile (node:internal/modules/cjs/loader:1191:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1245:10)
    at Module.load (node:internal/modules/cjs/loader:1069:32) {
  code: 'MODULE_NOT_FOUND',
  path: '/home/daniel/Rpanion-server/node_modules/tryer/package.json',
  requestPath: 'tryer'
}

```

**Comment by lida2003** [comment]
- 2023-02-21-raspios-bullseye-armhf-lite.img.xz + PiZeroW ==> Fail

```
+ cd ../../deploy
+ ./build_rpanion.sh
+ cd ../
++ free -m
++ awk '/^Mem:/{print $2}'
+ '[' 429 -le 500 ']'
+ export NODE_OPTIONS=--max-old-space-size=256
+ NODE_OPTIONS=--max-old-space-size=256
+ MAX_RETRIES=3
+ NPM_INSTALL_EXIT_CODE=0
++ seq 1 3
+ for i in $(seq 1 $MAX_RETRIES)
+ echo 'Running npm install (attempt 1)...'
Running npm install (attempt 1)...
+ npm install
npm WARN deprecated w3c-hr-time@1.0.2: Use your platform's native performance.now() and performance.timeOrigin.
npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
npm WARN deprecated rollup-plugin-terser@7.0.2: This package has been deprecated and is no longer maintained. Please use @rollup/plugin-terser
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
npm WARN deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated express-pino-logger@7.0.0: use pino-http instead
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm notice
npm notice New major version of npm available! 8.19.3 -> 10.3.0
npm notice Changelog: https://github.com/npm/cli/releases/tag/v10.3.0
npm notice Run npm install -g npm@10.3.0 to update!
npm notice
npm ERR! code ETIMEDOUT
npm ERR! syscall connect
npm ERR! errno ETIMEDOUT
npm ERR! network request to https://registry.npmjs.org/yn/-/yn-3.1.1.tgz failed, reason: connect ETIMEDOUT 104.16.0.35:443
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/daniel/.npm/_logs/2024-01-18T05_23_19_824Z-debug-0.log

```

**Comment by lida2003** [comment]
I think  Re-install process will fail due to the existence of binary(node/npm) links #194  is OK for the issue.
